core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -516,7 +516,15 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
     newProps ++= staticBrokerConfigs
     overrideProps(newProps, dynamicDefaultConfigs)
     overrideProps(newProps, dynamicBrokerConfigs)
-    val oldConfig = currentConfig
+
+    // We need a copy of the current config since `currentConfig` is initialized with `kafkaConfig`
+    // which means the call to `updateCurrentConfig` would end up mutating `oldConfig`.
+    val oldConfig = if (kafkaConfig eq currentConfig) {
```

Comment:
@hachikuji Thanks for the PR. Do you know if there is an issue only with KRraft or is there an issue with ZK as well? 

With ZK, the sequence is:
1. Create KafkaConfig with static configs from server.properties
2. Initialize KafkaConfig with initial configs from ZooKeeper using `DynamicBrokerConfig.initialize(zkClient)`. This always creates a new KafkaConfig, so we don't need this check?
3. Start DynamicConfigManager. All ZK updates after 2) are handled through change notifications. 
  
With KRaft, there doesn't seem to be an initialize() for initializing the state from existing dynamic configs, so are all configs handled similar to 3)? In which case, we should perhaps move the initialization of `currentConfig` from `DynamicBrokerConfig.initialize(zkClient)` to somewhere common for KRaft and avoid this check here?

Reply:
@rajinisivaram Hmm, good point. It looks like the bug only affects KRaft since there is no call to `initialize`. The code still feels a little slippery though, so maybe there is room for some defensiveness. Let me take a look.

Reply:
I pushed a patch which initializes `currentConfig` as null in order to make the call to `initialize` required. Let me know if that seems like a reasonable approach.